src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzerTests.GuardedCallsTests.cs

```diff
@@ -0,0 +1,1778 @@
+using System.Runtime.Versioning;
+using System;
+
+public class Test
+{
+    public void M1()
+    {
+        if(OperatingSystemHelper.IsOSPlatformVersionAtLeast(""Windows"", 10, 2))
+        {
+            OsDependentClass odc = new OsDependentClass();
+            odc.Method2();
+        }
+        else
+        {
+            OsDependentClass odc2 = [|new OsDependentClass()|];
+            [|odc2.Method2()|];
+        }
+    }
+}
+[SupportedOSPlatform(""Windows10.1.2.3"")]
+public class OsDependentClass
+{
+    public void Method2()
+    {
+    }
+}
+" + MockAttributesCsSource + MockRuntimeApiSource;
+            await VerifyAnalyzerAsyncCs(source);
+
+            var vbSource = @"
+Imports System.Runtime.Versioning
+Imports System
+
+Public Class Test
+    Public Sub M1()
+        If OperatingSystemHelper.IsOSPlatformVersionAtLeast(""Windows"", 10, 2) Then
+            Dim odc As OsDependentClass = New OsDependentClass()
+            odc.M2()
+        Else
+            Dim odc2 As OsDependentClass = [|New OsDependentClass()|]
+            [|odc2.M2()|]
+        End If
+    End Sub
+End Class
+
+<SupportedOSPlatform(""Windows10.1.2.3"")>
+Public Class OsDependentClass
+    Public Sub M2()
+    End Sub
+End Class
+" + MockRuntimeApiSourceVb + MockAttributesVbSource;
+            await VerifyAnalyzerAsyncVb(vbSource);
+        }
+
+        [Fact]
+        public async Task LocalFunctionCallsOsDependentMember_GuardedCall_SimpleIfElse()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+public class Test
+{
+    public void M1()
+    {
+        void Test()
+        {
+            if(OperatingSystemHelper.IsOSPlatformVersionAtLeast(""Windows"", 10, 2, 1))
+            {
+                M2();
+            }
+            else
+            {
+                [|M2()|];
+            }
+        }
+        Test();
+    }
+
+    [SupportedOSPlatform(""Windows10.1.2.3"")]
+    public void M2()
+    {
+    }
+}
+" + MockAttributesCsSource + MockRuntimeApiSource;
+            await VerifyAnalyzerAsyncCs(source);
+        }
+
+        [Fact]
+        public async Task LocalFunctionCallsPlatformDependentMember_InvokedFromDifferentContext()
+        {
+            var source = @"
+using System.Runtime.Versioning;
+using System;
+
+public class Test
+{
+    void M()
+    {
+        if (OperatingSystemHelper.IsOSPlatformVersionAtLeast(""Windows"", 10, 2))
+        {
+            LocalM(); // We don't need to account this platform check
+        }
+
+        LocalM();
+        return;
+
+        void LocalM()
+        {
+            if (OperatingSystemHelper.IsOSPlatformVersionAtLeast(""Windows"", 10, 2))
+            {
+                WindowsOnlyMethod();
+            }
+            else
+            {
+                [|WindowsOnlyMethod()|];
+            }
+            if (OperatingSystemHelper.IsWindows() && !OperatingSystemHelper.IsWindowsVersionAtLeast(10,0))
+            {
+                UnsupportedWindows10();
```

Comment:
This check is failing:

```
Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of test state
    Mismatch between number of diagnostics returned, expected "2" actual "1"
    
    Diagnostics:
    // /0/Test0.cs(33,17): warning CA1416: 'UnsupportedWindows10' is not supported or has been removed since Windows 10.0 version
    VerifyCS.Diagnostic(PlatformCompatabilityAnalyzer.CA1416).WithSpan(33, 17, 33, 39).WithArguments("UnsupportedWindows10", "Windows", "10.0"), 
```

Is it because of the negated check `!OperatingSystemHelper`?

Reply:
Weird; it looks right to me, and I've set up a repro of this in a local demo project and it's behaving consistently with what the test is asserting.

Reply:
Update: Actually, the code inside the `else` is what's failing (line 798), even though the diagnostic is expected.

Reply:
I set up a local repro of this and narrowed it down a bit.  It's weird.

With the following, no diagnostics show, but I would expect to see diagnostics on both calls to `WindowsOnlyMethod`.

```csharp
using System;
using System.Runtime.Versioning;

[assembly:SupportedOSPlatform("windows7.0")]

static class Program
{
    public static void Main()
    {
        if (OperatingSystem.IsWindowsVersionAtLeast(10))
        {
            WindowsSpecificApis.WindowsOnlyMethod();
        }
        else
        {
            WindowsSpecificApis.WindowsOnlyMethod();
        }
    }
}

public class WindowsSpecificApis
{
    [SupportedOSPlatform("windows10.1.2.3")]
    public static void WindowsOnlyMethod() { }

    [UnsupportedOSPlatform("windows10.1.2.3")]
    public static void UnsupportedWindows10() { }
}
```

Strangely, if I comment out the `[assembly: SupportedOSPlatform("windows7.0")]` attribute, then I do get diagnostics on both calls.  This behavior seems unique to the use of `[SupportedOSPlatform]`; the behavior seems to be functioning properly for `[UnsupportedOSPlatform]`.

Reply:
It looks like it might be the same issue that Immo filed here: https://github.com/dotnet/roslyn-analyzers/issues/4009

Reply:
> Do you know why?

Yes i know, it is related to https://github.com/dotnet/roslyn-analyzers/pull/3917#discussion_r470222550
It is failing because the first call of `LocalM()` is within guarded call, second is not. as @mavasani mentioned i should have check all elements within interprocedural result instead of using only 1st one. It is failing randomly depending on which of the `LocalM()` calls came out at first if guarded one comes it fails if 2nd unguarded one comes it passes. 
1. I have divided the test into two as we cannot test this scenario completely within one method
2. I have been looking into it in the afternoon and couldn't figure out how to detect the invocation for evaluation, now waiting what @mavasani suggest

Reply:
@jeffhandley it is not the same issue you mentioned, i will look at your example

Reply:
@jeffhandley your example was indeed same as #4009 and now fixed, the test [added ](https://github.com/dotnet/roslyn-analyzers/pull/3917/commits/28a7cba35522a24494a41dc66f6b12e52d173512#diff-143b956da2811185e4067197cdaed6fcR333-R366)